File: benchmarks/byte-scan-view/byte_scan_view.rs
// Rust independent baseline for BMD3 byte-scan-view.
// The program prints no output until the bounded input is fully validated.

use std::env;
use std::fs::File;
use std::io::{self, Read};

const MAX_INPUT_BYTES: u64 = 64 * 1024 * 1024;

fn parse_delimiter(text: &str) -> Option<u8> {
    if text.is_empty() || matches!(text.as_bytes().first(), Some(b'+' | b'-')) ||
        (text.as_bytes().first() == Some(&b'0') && text.len() != 1) {
        return None;
    }
    let value: u16 = text.parse().ok()?;
    u8::try_from(value).ok()
}

fn run(path: &str, delimiter: u8) -> io::Result<(u64, u64)> {
    let mut input = File::open(path)?;
    let mut buffer = [0_u8; 64 * 1024];
    let mut bytes = 0_u64;
    let mut matches = 0_u64;

    loop {
        let count = input.read(&mut buffer)?;
        if count == 0 {
            break;
        }
        let count = u64::try_from(count).expect("buffer length fits in u64");
        if bytes > MAX_INPUT_BYTES - count {
            return Err(io::Error::new(io::ErrorKind::InvalidData, "input exceeds 64 MiB"));
        }
        bytes += count;
        matches += buffer[..usize::try_from(count).expect("buffer length fits in usize")]
            .iter()
            .filter(|&&byte| byte == delimiter)
            .count() as u64;
    }
    Ok((bytes, matches))
}

fn main() {
    let arguments: Vec<String> = env::args().collect();
    if arguments.len() != 3 {
        std::process::exit(2);
    }
    let delimiter = match parse_delimiter(&arguments[2]) {
        Some(value) => value,
        None => std::process::exit(2),
    };
    let (bytes, matches) = match run(&arguments[1], delimiter) {
        Ok(value) => value,
        Err(_) => std::process::exit(3),
    };
    print!("{{\"bytes\":\"{}\",\"matches\":\"{}\"}}", bytes, matches);
}

File: benchmarks/byte-scan-view/byte_scan_view.c
// C23 independent baseline for BMD3 byte-scan-view.
// The program prints no output until the bounded input is fully validated.

#include <errno.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum { MAX_INPUT_BYTES = 64 * 1024 * 1024, BUFFER_BYTES = 64 * 1024 };

static int parse_delimiter(const char *text, uint8_t *delimiter) {
  char *end = NULL;
  unsigned long value;

  if (text == NULL || text[0] == '\0' || text[0] == '+' || text[0] == '-' ||
      (text[0] == '0' && text[1] != '\0')) return 0;
  errno = 0;
  value = strtoul(text, &end, 10);
  if (errno != 0 || end == text || *end != '\0' || value > 255UL) return 0;
  *delimiter = (uint8_t)value;
  return 1;
}

int main(int argc, char **argv) {
  FILE *input;
  uint8_t delimiter;
  unsigned char buffer[BUFFER_BYTES];
  uint64_t bytes = 0;
  uint64_t matches = 0;
  size_t count;

  if (argc != 3 || !parse_delimiter(argv[2], &delimiter)) return 2;
  input = fopen(argv[1], "rb");
  if (input == NULL) return 3;

  while ((count = fread(buffer, 1, sizeof(buffer), input)) != 0) {
    size_t index;
    if (bytes > (uint64_t)MAX_INPUT_BYTES - (uint64_t)count) {
      fclose(input);
      return 4;
    }
    bytes += (uint64_t)count;
    for (index = 0; index < count; ++index) {
      if (buffer[index] == delimiter) ++matches;
    }
  }
  if (ferror(input) || fclose(input) != 0) return 5;

  if (printf("{\"bytes\":\"%" PRIu64 "\",\"matches\":\"%" PRIu64 "\"}", bytes, matches) < 0) return 6;
  return 0;
}
